realm/realm-library/src/main/java/io/realm/RealmList.java

```diff
@@ -218,15 +218,30 @@ public E set(int location, E object) {
 
     // Transparently copies a standalone object or managed object from another Realm to the Realm backing this RealmList.
     private E copyToRealmIfNeeded(E object) {
-        // Object is already in this realm
-        if (object.row != null && object.realm.getPath().equals(realm.getPath())) {
-            return object;
-        }
-
-        // We don't support moving DynamicRealmObjects across Realms automatically. The overhead is too big as you
-        // have to run a full schema validation for each object.
         if (object instanceof DynamicRealmObject) {
-            throw new IllegalArgumentException("Automatically copying DynamicRealmObjects from other Realms are not supported");
+            String listClassName = view.getTargetTable().getName();
+            String objectClassName = ((DynamicRealmObject) object).row.getTable().getName();
+            if (object.realm == realm && listClassName.equals(objectClassName)) {
+                // Same Realm instance and same target table
+                return object;
+            } else if (object.realm == realm) {
```

Comment:
It looks a bit funny with 3 these if's at the same level, I would probably inline this in the first if like this:
```
if (object.realm == realm) {
  if (listClassName.equals(objectClassName) {
   //...
   } else {
   //...
   }
} else {

}
```